core/finality-grandpa/src/lib.rs

```diff
@@ -0,0 +1,767 @@
+    type Error = Error;
+
+    fn round_data(
+        &self,
+        round: u64
+    ) -> voter::RoundData<Self::Timer, Self::Id, Self::In, Self::Out> {
+		use client::BlockchainEvents;
+		use tokio::timer::Delay;
+
+		let now = Instant::now();
+		let prevote_timer = Delay::new(now + self.config.gossip_duration * 2);
+		let precommit_timer = Delay::new(now + self.config.gossip_duration * 4);
+
+		// TODO: dispatch this with `mpsc::spawn`.
+		let incoming = checked_message_stream::<Block, _>(
+			self.network.messages_for(round),
+			self.config.voters.clone(),
+		);
+
+		let (out_rx, outgoing) = outgoing_messages::<Block, _>(
+			self.config.local_key.clone(),
+			self.config.voters.clone(),
+			round,
+			self.network.clone(),
+		);
+
+		// schedule incoming messages from the network to be held until
+		// corresponding blocks are imported.
+		let incoming = UntilImported::new(
+			self.inner.import_notification_stream(),
+			self.inner.clone(),
+			incoming,
+		);
+
+		// join incoming network messages with locally originating ones.
+		let incoming = Box::new(incoming.select(out_rx));
+
+		// schedule network message cleanup when sink drops.
+		let outgoing = Box::new(ClearOnDrop {
+			round,
+			network: self.network.clone(),
+			inner: outgoing,
+		});
+
+		voter::RoundData {
+			prevote_timer: Box::new(prevote_timer.map_err(Error::Timer)),
+    		precommit_timer: Box::new(precommit_timer.map_err(Error::Timer)),
+    		voters: self.voters.clone(),
+    		incoming,
+    		outgoing,
+		}
+	}
+
+    fn completed(&self, round: u64, state: RoundState<Block::Hash>) {
+		let encoded_state = (round, state).encode();
+		if let Err(e) = self.inner.backend()
+			.insert_aux(&[(LAST_COMPLETED_KEY, &encoded_state[..])], &[])
+		{
+			warn!(target: "afg", "Error bookkeeping last completed round in DB: {:?}", e);
+		}
+	}
+
+    fn finalize_block(&self, hash: Block::Hash, number: u32) {
+		// TODO: don't unconditionally notify.
+		if let Err(e) = self.inner.finalize_block(BlockId::Hash(hash), true) {
+			warn!(target: "afg", "Error applying finality to block {:?}: {:?}", (hash, number), e);
+		}
+	}
+
+    fn prevote_equivocation(
+        &self,
+        _round: u64,
+        equivocation: ::grandpa::Equivocation<Self::Id, Prevote<Block::Hash>, Self::Signature>
+    ) {
+		warn!(target: "afg", "Detected prevote equivocation in the finality worker: {:?}", equivocation);
+		// nothing yet; this could craft misbehavior reports of some kind.
+	}
+
+    fn precommit_equivocation(
+        &self,
+        _round: u64,
+        equivocation: Equivocation<Self::Id, Precommit<Block::Hash>, Self::Signature>
+    ) {
+		warn!(target: "afg", "Detected precommit equivocation in the finality worker: {:?}", equivocation);
+		// nothing yet
+	}
+}
+
+/// Run a GRANDPA voter as a task. The returned future should be executed in a tokio runtime.
+pub fn run_voter<B, E, Block: BlockT, N>(
```

Comment:
Isn't this the `Voter` we can/should run in order to just observe the `GRANDPA`-Process taking place without actually voting ourselves? I understand in that case I'd just provide a dummy `Network`, which isn't actually sending messages, but giving us information about potential misbehavior we could flag, but I feel like the function name and docs might already want to anticipate that, don't you think? Something more generic like `run_grandpa` instead (and maybe add a TODO to the docs to add example usage of both cases?). What do you think?

Reply:
this is a voter that can send messages and cast votes. it can be a non-active voter. changing it to `run_grandpa` is uncontroversial to me